tokens/src/spl_token.rs

```diff
@@ -0,0 +1,698 @@
+        .await?
+        .unwrap_or_default();
+    let source_token = SplTokenAccount::unpack(&source_token_account.data)?;
+    if source_token.amount < allocation_amount {
+        return Err(Error::InsufficientFunds(
+            vec![FundingSource::SplTokenAccount].into(),
+            token_amount_to_ui_amount(allocation_amount, spl_token_args.decimals).ui_amount,
+        ));
+    }
+    Ok(())
+}
+
+pub async fn print_token_balances(
+    client: &mut BanksClient,
+    allocation: &Allocation,
+    spl_token_args: &SplTokenArgs,
+) -> Result<(), Error> {
+    let address = allocation.recipient.parse().unwrap();
+    let expected = allocation.amount;
+    let associated_token_address = get_associated_token_address(
+        &spl_token_v2_0_pubkey(&address),
+        &spl_token_v2_0_pubkey(&spl_token_args.mint),
+    );
+    let recipient_account = client
+        .get_account(pubkey_from_spl_token_v2_0(&associated_token_address))
+        .await?
+        .unwrap_or_default();
+    let (actual, difference) =
+        if let Ok(recipient_token) = SplTokenAccount::unpack(&recipient_account.data) {
+            let actual = token_amount_to_ui_amount(recipient_token.amount, spl_token_args.decimals)
+                .ui_amount;
+            (
+                style(format!(
+                    "{:>24.1$}",
+                    actual, spl_token_args.decimals as usize
+                )),
+                format!(
+                    "{:>24.1$}",
+                    actual - expected,
+                    spl_token_args.decimals as usize
+                ),
+            )
+        } else {
+            (
+                style("Associated token account not yet created".to_string()).yellow(),
+                "".to_string(),
+            )
+        };
+    println!(
+        "{:<44}  {:>24.4$}  {:>24}  {:>24}",
+        allocation.recipient, expected, actual, difference, spl_token_args.decimals as usize
+    );
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        commands::{process_allocations, tests::tmp_file_path, Allocation},
+        db::{self, check_output_file},
+        spl_token::spl_token_amount,
+    };
+    use solana_account_decoder::parse_token::{spl_token_id_v2_0, spl_token_v2_0_pubkey};
+    use solana_program_test::*;
+    use solana_sdk::{
+        hash::Hash,
+        signature::{read_keypair_file, write_keypair_file, Keypair, Signer},
+        system_instruction,
+        transaction::Transaction,
+    };
+    use solana_transaction_status::parse_token::spl_token_v2_0_instruction;
+    use spl_associated_token_account_v1_0::{
+        create_associated_token_account, get_associated_token_address,
+    };
+    use spl_token_v2_0::{
+        instruction::{initialize_account, initialize_mint, mint_to},
+        solana_program::pubkey::Pubkey,
+    };
+    use tempfile::{tempdir, NamedTempFile};
+
+    fn program_test() -> ProgramTest {
```

Comment:
I'm glad that ProgramTest is serving a purpose in the monorepo as well!  I wasn't expecting that